### DISCUSSION TITLE
ESP32-S2 change USB Midi Device name

### DISCUSSION THREAD

**Discussion by Vahen1981** [tttapa/Control-Surface#1065] [Q&A]
Hi everyone!
After some hours working I finally could use the ESP32-S2 as a midi usb device, but I can't change the name of the USB midi device, so my system recognizes with the name of "TinyUSB MIDI".
Someone knows how to handle this?

**Comment by tttapa** [comment]
I can't test it right now, but IIRC, you can use `USB.productName("foo")`:

https://github.com/espressif/arduino-esp32/blob/cd3d0bf569f8c0ec1d9ed549c649b9d237d10e28/cores/esp32/USB.h#L82

**Comment by Vahen1981** [comment]
Thank you Pieter! It compiles, but the name doesn't change :(

**Comment by Vahen1981** [comment]
Hi, I finally did it... I found the default name in the ESP32 core and I change it... if it is usefull for someone is the file USBMIDI.CPP, located in User\AppData\Local\Arduino15\packages\esp32\hardware\esp32\3.0.4\libraries\USB\src\
the line 21 has the default name of the device.

**Reply by micdre** [reply]
Hi , i was looking to change the device name too. Did what you said, but it still shows 'TinyUSB MIDI' as de device name in Windows. Any ideas about this? Is there another file with the name in it?

**Reply by Vahen1981** [reply]
I'm not sure, but it might be related to Windows. You could try completely uninstalling the device and “cleaning” the COM ports on your computer.

**Comment by AndyBodnar** [comment]
To change the USB MIDI device name on ESP32-S2, you need to modify the TinyUSB descriptor before initialization.

**Method 1: Arduino IDE**
Add these defines BEFORE including any headers:
```cpp
#define PRODUCT_NAME "My Custom MIDI"
#define MANUFACTURER_NAME "My Company"
```

**Method 2: Modify tusb_config.h**
Find your TinyUSB config and change:
```c
#define CFG_TUD_MIDI_PRODUCT_STR "My Custom MIDI Device"
```

**Method 3: Runtime descriptor (Control-Surface specific)**
If using Control-Surface library, check if theres a setName() or similar before begin().

**Important:** After changing, you may need to:
1. Unplug and replug the USB
2. Clear device cache on your OS (Windows especially caches USB names)
3. On Mac: delete /Library/Preferences/com.apple.audio.DeviceSettings.plist

The name TinyUSB MIDI comes from the default TinyUSB descriptor, so youre on the right track - just need to override it before USB init.

**Reply by tttapa** [reply]
This response reads like it was generated by an LLM. Could you report which of the three methods worked for you?
